src/nupic/algorithms/ColumnPooler.cpp

```diff
@@ -0,0 +1,565 @@
+ * ----------------------------------------------------------------------
+ */
+
+/** @file
+ * Implementation of ColumnPooler
+ */
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <iomanip> // std::setprecision
+#include <functional> // function
+
+// #include <nupic/algorithms/ColumnPooler.hpp>
+#include <nupic/types/Types.hpp>
+#include <nupic/types/Serializable.hpp>
+#include <nupic/types/Sdr.hpp>
+#include <nupic/utils/SdrMetrics.hpp>
+#include <nupic/algorithms/Connections.hpp>
+#include <nupic/algorithms/TemporalMemory.hpp>
+#include <nupic/math/Math.hpp>
+#include <nupic/math/Topology.hpp>
+
+namespace nupic {
+namespace algorithms {
+namespace column_pooler {
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::math::topology;
+using namespace nupic::algorithms::connections;
+using nupic::algorithms::temporal_memory::TemporalMemory;
+
+
+// TODO: Connections learning rules are different.
+
+
+/* Topology( location, potentialPool ) */
+typedef function<void(SDR&, SDR&)> Topology_t;
+
+class ColumnPooler // : public Serializable
+{
+private:
+  vector<UInt> proximalInputDimensions_;
+  vector<UInt> distalInputDimensions_;
+  vector<UInt> inhibitionDimensions_;
+  vector<UInt> cellDimensions_;
+  UInt         cellsPerInhibitionArea_;
+  UInt         proximalSegments_;
+
+  vector<UInt32> rawOverlaps_;
+  vector<UInt> proximalMaxSegment_;
+
+  Random rng_;
+  vector<Real> tieBreaker_;
+  UInt iterationNum_;
+  UInt iterationLearnNum_;
+
+public:
+  const vector<UInt> &proximalInputDimensions = proximalInputDimensions_;
+  const vector<UInt> &distalInputDimensions   = distalInputDimensions_;
+  const vector<UInt> &inhibitionDimensions    = inhibitionDimensions_;
+  const UInt         &cellsPerInhibitionArea  = cellsPerInhibitionArea_;
+  const vector<UInt> &cellDimensions          = cellDimensions_;
+  const UInt         &proximalSegments        = proximalSegments_;
+
+  Real sparsity;
+
+  UInt       proximalSegmentThreshold;
+  Permanence proximalIncrement;
+  Permanence proximalDecrement;
+  Permanence proximalSynapseThreshold;
+
+  SDR_ActivationFrequency *AF;
+  Real stability_rate;
+  Real fatigue_rate;
+  vector<Real> X_act;
+  vector<Real> X_inact;
+
+  const UInt &iterationNum      = iterationNum_;
+  const UInt &iterationLearnNum = iterationLearnNum_;
+
+  /**
+   * The proximal connections have regular structure.  Cells in an inhibition
+   * area are contiguous and all segments on a cell are contiguous.  This
+   * allows fast index math instead of slowers lists of ID's.
+   */
+  Connections proximalConnections;
+
+  TemporalMemory distalConnections;
+
+
+  ColumnPooler() {}; //default constructor, must call initialize to setup properly
+ 
+  ColumnPooler(
+    const vector<UInt> proximalInputDimensions,
+    const vector<UInt> distalInputDimensions,
+    const vector<UInt> inhibitionDimensions,
+    UInt               cellsPerInhibitionArea,
```

Comment:
provide sensible defaults for the constructor variables. (MNIST example a good choice?)